modin/engines/base/frame/data.py

```diff
@@ -976,37 +976,83 @@ def _join_index_objects(self, axis, other_index, how, sort):
 
         Parameters
         ----------
-            axis : 0 or 1
-                The axis index object to join (0 - rows, 1 - columns).
-            other_index : Index
-                The other_index to join on.
-            how : {'left', 'right', 'inner', 'outer'}
-                The type of join to join to make.
-            sort : boolean
-                Whether or not to sort the joined index
+        axis : 0 or 1
+            The axis index object to join (0 - rows, 1 - columns).
+        others_index : list(Index)
+            The others_index to join on. Index of `self` frame should be first.
+        how : {'left', 'right', 'inner', 'outer'}
+            The type of join to join to make.
+        sort : boolean
+            Whether or not to sort the joined index
 
         Returns
         -------
-        Index
-            Joined indices.
+        (Index, func)
+            Joined index with make_reindexer func
         """
+        assert isinstance(others_index, list)
 
-        def merge_index(obj1, obj2):
+        # define helper functions
+        def merge(left_index, right_index):
             if axis == 1 and how == "outer" and not sort:
-                return obj1.union(obj2, sort=False)
+                return left_index.union(right_index, sort=False)
             else:
-                return obj1.join(obj2, how=how, sort=sort)
-
-        if isinstance(other_index, list):
-            joined_obj = self.columns if axis else self.index
-            # TODO: revisit for performance
-            for obj in other_index:
-                joined_obj = merge_index(joined_obj, obj)
-            return joined_obj
-        if axis:
-            return merge_index(self.columns, other_index)
+                return left_index.join(right_index, how=how, sort=sort)
+
+        # define condition for joining indexes
+        self_index = others_index[0]
+        do_join_index = False
+        for index in others_index[1:]:
+            if not self_index.equals(index):
+                do_join_index = True
+                break
+
+        # define condition for joining indexes with getting indexersы
+        is_duplicates = (
+            any([not index.is_unique for index in others_index]) and axis == 0
```

Comment:
it's better to pass generator object to `any` instead of a list:
```python
any(condition(x) for x in some_list) # generator
any([condition(x) for x in some_list]) # list
```
when we're working with a list we should instantiate the whole of it, so calculate the condition for all elements in `some_list` and after that call `any` on the resulted list. When we're working with a generator we don't need to instantiate the whole sequence, after the first `True` condition `any` can consider that it should return `True`.

Check this example:
```python
>>> def condition(x):
...     print(f"called on {x}")
...     return True if x >= 1 else False
...
>>> any(condition(x) for x in range(5))
called on 0
called on 1
True
>>> any([condition(x) for x in range(5)])
called on 0
called on 1
called on 2
called on 3
called on 4
True
```

Reply:
@dchigarev good decision! :)